[TableGen] Use bitwise operations to access HwMode ID.

superZWT123 opened the pull request:
1. Bitwise operations are used to access HwMode, allowing for the coexistence of HwMode IDs for different features (such as RegInfo and EncodingInfo). This will provide better scalability for HwMode. Currently, most users utilize HwMode primarily for configuring  Register-related information, and few use it for configuring Encoding. The limited scalability of HwMode has been a significant factor in this usage pattern.
2. Sink the HwMode Encodings selection logic down to per instruction level, this makes the logic for choosing encodings clearer and provides better error messages.
3. Add some HwMode ID conflict detection to the getHwMode() interface.

Commit:
[TableGen] Use bitwise operations to access HwMode ID.

1. Bitwise operations are used to access HwMode, allowing for the
   coexistence of HwMode IDs for different features (such as RegInfo
   and EncodingInfo). This will provide better scalability for HwMode.
   Currently, most users utilize HwMode primarily for configuring
   Register-related information, and few use it for configuring Encoding.
   The limited scalability of HwMode has been a significant factor in this
   usage pattern.
2. Sink the HwMode Encodings selection logic down to per instruction level,
   this makes the logic for choosing encodings clearer and provides better
   error messages.
3. Add some HwMode ID conflict detection to the getHwMode() interface.

superZWT123 commented:
[PRE-MR1](https://github.com/llvm/llvm-project/pull/86060)

superZWT123 commented:
@wangpc-pp @nvjle @topperc @kparzysz , hi  if any of you could help me review this PR, I would be greatly thankful~

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1150,15 +1150,21 @@ def Feature32Bit
     : SubtargetFeature<"32bit", "IsRV32", "true", "Implements RV32">;
 def Feature64Bit
     : SubtargetFeature<"64bit", "IsRV64", "true", "Implements RV64">;
+def FeatureEncodingTmp
```
What's this for? Testing?

superZWT123 replied:
Yes, this is just a reference and will be deleted eventually. It’s mainly to help everyone easily understand the purpose of this patch.

superZWT123 commented:
Ping

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoZcmop.td:
```diff
@@ -11,22 +11,60 @@
 //
 //===----------------------------------------------------------------------===//
 
+multiclass tmpEclass<bits<3> imm3> {
```
Please do not add temporary code to a real target. You should write a unit test for your change. There many in llvm/test/TableGen/ already.

superZWT123 replied:
Sure, in this PR, only the first commit is intended for merging into the main branch. The second commit serves merely as a helpful reference for the review process and is not meant for merging. I can promptly remove it if necessary. I provided a reference implementation on RISCV backend because I think that the testcases in the /TableGen directory may not adequately showcase the PR’s impact fully. Thank you.

nvjle commented:
Though I have only briefly reviewed this (beyond what I did in the original series a few weeks ago), we did apply the patches locally to a production downstream back end. Unfortunately, we're getting a number of failures in both disassembler and encoding tests. 

Just a quick diff of the `MyTargetMCCodeEmitter.inc` shows ~15000 differences compared to pre-patch. It is impossible to tell immediately what is incorrect on such a complicated target/ISA/multiple-encodings, and we're busy on other high priority work at the moment. I/we will attempt to isolate this in the next few days, but it may be that we are unable with current resources in a short amount of time. If this is committed as is, our target (and possibly others downstream) will be broken on the next upstream rebase. Maybe it is something easy, we will try our best to figure it out.

One immediate hunch is that our back end has long instruction words of a few different widths-- starting at 128 bits and only getting larger from there. This means that the decoder and encoder emitters switch to APInt mode, which may or may not have been carefully tested in your patch. No in-tree back ends have long words, but at least a few downstream ones do-- and I'm not sure if yours does. 

As an aside, it is difficult to avoid breakage in different downstream back ends for features that are not (yet) used in any in-tree back ends. Ours will be upstreamed eventually, but the timeline is unknown. Until then, it is possible changes like this in one or another downstream target will need to stay local (we have many).

Also, I agree with Craig that the pipe-cleaning/test changes in the RISCV back end need to be removed. They also lengthen and obscure the core patch. The unit tests should cover all of that functionality. If possible, please also make sure the long-word APInt modes of the encoder (and anywhere else changes were made) are covered in the testing. It isn't possible, of course, to do actual disassembler/encoder tests since there is no in-tree back end, but at least you can sanity check the encoder emitter output for that mode (I didn't see it in the current tests).

nvjle commented on llvm/test/TableGen/HwModeEncodeDecode3.td:
```diff
@@ -178,18 +185,53 @@ def unrelated: Instruction {
-
-// ENCODER:     case ::foo: {
-// ENCODER:      switch (HwMode) {
-// ENCODER:      default: llvm_unreachable("Unhandled HwMode");
```
Please add one more hardware mode `ModeC` and use it in one of the `EncodingInfos` (or add a new unit test file altogether). The reason why will be clear later in this review where I discuss the emitter code that generates the above-- and it would only show if there is at least one more mode.

nvjle commented on llvm/test/TableGen/HwModeEncodeDecode3.td:
```diff
@@ -160,15 +160,22 @@ def unrelated: Instruction {
 // ENCODER-LABEL:   static const uint64_t InstBits_ModeA[] = {
 // ENCODER:         UINT64_C(2),        // bar
-// ENCODER:         UINT64_C(0),        // baz
+// ENCODER:         UINT64_C(12),       // baz
```
It doesn't make sense to me to assign the encoding of a different mode for these unused entries. Why don't we leave them as zeroes as they originally were? In theory it doesn't matter, but it seems confusing to assign a non-zero value from some other mode. Also, for downstream back ends with any failures, diff'ing the original and new emitted files will have fewer "irrelevant" differences.

nvjle commented on llvm/utils/TableGen/CodeEmitterGen.cpp:
```diff
@@ -290,11 +290,55 @@ CodeEmitterGen::getInstructionCases(Record *R, CodeGenTarget &Target) {
+          append("      case " + itostr(DefaultMode) +
+                 ": InstBitsByHw = InstBits");
+        } else {
+          append("      case " + itostr(1 << (ModeId - 1)) +
```
As mentioned earlier, I think there is a bug here with generating these case values. That is, the switch is on `HwMode`,which was assigned by the `getHwMode()` above it. Now as I understand it, `getHwMode` returns a HwMode integer value corresponding to a particular mode-- not a bit position in the bit vector. On the other hand, the case label is `1 << (ModeId - 1)`-- a bit position. It seems that we're comparing apples and oranges. Your test case wouldn't show this because the expressions happen to be equal for 0, 1, and 2. If you add another mode, I think you'll find it's case value is 4, not 3.

That would certainly explain at least some of the failures we're seeing downstream where we do indeed will see a  `ModeId` larger than 2.

superZWT123 replied:
Thank you very much, I believe this is currently the biggest issue. There is indeed a logical problem here that I will fix shortly. It’s due to changes I made to the interface, which resulted in mishandling the logic here. Thank you very much!

nvjle commented on llvm/utils/TableGen/CodeEmitterGen.cpp:
```diff
@@ -290,11 +290,55 @@ CodeEmitterGen::getInstructionCases(Record *R, CodeGenTarget &Target) {
+        if (ModeId == DefaultMode) {
+          append("      case " + itostr(DefaultMode) + ": {\n");
+        } else {
+          append("      case " + itostr(1 << (ModeId - 1)) + ": {\n");
```
Same issue here.

nvjle commented on llvm/utils/TableGen/CodeEmitterGen.cpp:
```diff
@@ -383,8 +427,17 @@ void CodeEmitterGen::emitInstructionBaseValues(
+        } else {
+          // If this Instr dosen't have this HwMode, just choose
+          // the encoding from the first HwMode. Otherwise, the encoding
+          // info would be empty.
```
As mentioned in the comments regarding the lit test corresponding to this code, it is unclear to me why we would want to arbitrarily choose to assign an unused entry to the encoding for a different mode. Originally the value would ultimately end up zero, which seems to make more sense to me.

But even if this is kept, the comment is a bit garbled.

nvjle commented on llvm/utils/TableGen/CodeEmitterGen.cpp:
```diff
@@ -290,11 +290,55 @@ CodeEmitterGen::getInstructionCases(Record *R, CodeGenTarget &Target) {
+      }
+
+      // Get HwModes for this Instr by bitwise AND operations,
+      // and find the table to which this instr and hwmode belong.
```
Nit: Be consistent in the prose-- we use `Instr` here and then `instr` there, and `HwModes` here vs `hwmode` there. And just say "mask" rather than the long-winded "by bitwise AND oeprations".

 I'd probably just write "Get the applicable hardware modes for this instruction by masking off the relevant bits."

nvjle commented on llvm/utils/TableGen/SubtargetEmitter.cpp:
```diff
@@ -1781,13 +1781,68 @@ void SubtargetEmitter::EmitHwModeCheck(const std::string &ClassName,
 
-  OS << "unsigned " << ClassName << "::getHwMode() const {\n";
+  // Collect all HwModes and related features defined in the TD files,
+  // and store them in bit format.
```
Similarly to earlier comments: "...store them in bit format." -> "store them as a bit vector."

nvjle commented on llvm/utils/TableGen/SubtargetEmitter.cpp:
```diff
@@ -1781,13 +1781,68 @@ void SubtargetEmitter::EmitHwModeCheck(const std::string &ClassName,
+
+  // Start emitting for getHwModeSet().
+  OS << "unsigned " << ClassName << "::getHwModeSet() const {\n";
+  OS << "  // Collect HwModes and store them in bits\n";
```
Even though this is auto-generated code, we should try to adhere to the Coding Standards (i.e., comments are proper prose with punctuation, full stop, etc.).

nvjle commented on llvm/utils/TableGen/SubtargetEmitter.cpp:
```diff
@@ -1781,13 +1781,68 @@ void SubtargetEmitter::EmitHwModeCheck(const std::string &ClassName,
+  OS << "unsigned " << ClassName
+     << "::getHwMode(enum HwModeType type) const {\n";
+  OS << "  unsigned Modes = getHwModeSet();\n";
+  OS << "\n  if(!Modes)\n    return Modes;\n\n";
```
As above, CS nit: `if(!Modes)` -> `if (!Modes)`

nvjle commented on llvm/utils/TableGen/SubtargetEmitter.cpp:
```diff
@@ -1781,13 +1781,68 @@ void SubtargetEmitter::EmitHwModeCheck(const std::string &ClassName,
+     << "    if (!llvm::has_single_bit<unsigned>(Modes))\n"
+     << "      llvm_unreachable(\"Two or more HwModes for Encoding were "
+        "found!\");\n"
+     << "    return llvm::countr_zero(Modes) + 1;\n  }\n";
```
All three of the non-Default cases are essentially identical other than the mask value. We should collapse this into ~1/3 the code.

nvjle commented on llvm/utils/TableGen/SubtargetEmitter.cpp:
```diff
@@ -1781,13 +1781,68 @@ void SubtargetEmitter::EmitHwModeCheck(const std::string &ClassName,
+        "found!\");\n"
+     << "    return llvm::countr_zero(Modes) + 1;\n  }\n";
+  OS << "  }\n";
+  OS << "  return 0; // should not get here\n}\n";
```
Should this instead be `llvm_unreachable("unexpected HwModeType")`?

superZWT123 commented:
Hi @nvjle, Thank you very much for your review, I will respond as promptly as possible. My backend has 128-bit wide instructions, so my initial development was based on **APInt** mode. Later, the testing was switched to a 64-bit instruction backend, some issues were detected during the switching testing process, and I promptly fixed them. The functionality works correctly in both cases.

These days, I am also quite busy as the company has many development tasks that need my attention, so I can only allocate some spare time to modify this patch. Nonetheless, I am eager to contribute to the community and help drive improvements in this module.

As you mentioned, the changes in this patch are sensitive and have a wide impact, so there is no rush. I will proceed slowly to ensure the correctness of the features. Your review comments are crucial in this process, so once again, thank you~

Commit:
[TableGen] fix HwMode logic in CodeEmitterGen

1. Since we have obtained HwMode Id from getHwMode() interface,
   we no longer need to perform bit related operations when checking
   the case.
2. When generating the 'InstBits' table, if the instruction does not have the
   HwMode ID corresponding to the current hwmode table, the base encoding
   will be generated as '0'.
3. Add testcase to test the correctness of hwmode encoding in APInt mode
4. Fixed a bug in the 'getOperandBitOffset' interface: it would check the
   HwMode ID for a instruction but did not preemptively fetch the HwMode ID.
5. Fix up some comments and clean up some codes.

superZWT123 commented:
Hi @nvjle , I'm really sorry for the delay in addressing your review feedback. This update includes the following five main changes:

1. Fixed the issue with improper HwMode retrieval during Encoding generation process (the most critical issue).
2. Used encoding '0' to represent instructions with mismatched HwMode in the encoding table.
3. Added testcase to assess instruction encoding under APInt mode (128-bit length) and introduced more HwModes to validate performance under multiple HwMode scenarios.
4. Discovered and resolved a potential issue (not introduced by my patch) where the existing 'getOperandBitOffset' interface was checking HwMode ID without first retrieving it.
5. Made some modifications related to Clean Code and also fixed the comments.

Could you please review these changes again and verify them in your backend? Thank you very much!

superZWT123 commented:
Hi @iii-i ,  I have discovered and resolved a potential issue where the existing 'getOperandBitOffset' interface was checking HwMode ID without first retrieving it. Could you please help me review the implementation of that part? Thank you very much!

nvjle commented:
> Hi @nvjle , I'm really sorry for the delay in addressing your review feedback. This update includes the following five main changes:
> 
>     1. Fixed the issue with improper HwMode retrieval during Encoding generation process (the most critical issue).
> 
>     2. Used encoding '0' to represent instructions with mismatched HwMode in the encoding table.
> 
>     3. Added testcase to assess instruction encoding under APInt mode (128-bit length) and introduced more HwModes to validate performance under multiple HwMode scenarios.
> 
>     4. Discovered and resolved a potential issue (not introduced by my patch) where the existing 'getOperandBitOffset' interface was checking HwMode ID without first retrieving it.
> 
>     5. Made some modifications related to Clean Code and also fixed the comments.
> 
> 
> Could you please review these changes again and verify them in your backend? Thank you very much!

Please give me a few days to re-review this iteration and incorporate into our repo to verify failures are gone.

nvjle reviewed (approved):
Thank you @superZWT123, all concerns have been addressed-- and this now passes when incorporated to our downstream back end.

LGTM, ready to land.

iii-i commented:
The `getInstructionCases()` change looks good to me, thanks for handling the issue there! I most likely didn't notice it, because SystemZ, which is currently the only user of `getOperandBitOffset()`, does not use `EncodingInfos`.
